src/agent/src/device.rs

```diff
@@ -436,217 +426,242 @@ fn scan_scsi_bus(scsi_addr: &str) -> Result<()> {
+                    ));
+                }
+            }
+
+            if resupdates
+                .insert((&specdev.r#type, host_major, host_minor), update)
+                .is_some()
+            {
+                return Err(anyhow!(
+                    "Conflicting device information for host_major={} host_minor={}",
+                    host_major,
+                    host_minor
+                ));
+            }
+        }
     }
 
-    let meta = fs::metadata(vm_path)?;
-    let dev_id = meta.rdev();
-    unsafe {
-        major_id = major(dev_id);
-        minor_id = minor(dev_id);
+    // Make sure we applied all of our updates
+    if !updates.is_empty() {
+        return Err(anyhow!(
+            "Missing devices in OCI spec: {:?}",
+            updates.into_keys().collect::<Vec<&str>>()
+        ));
     }
 
-    info!(
-        sl!(),
-        "update_spec_device(): vm_path={}, major: {}, minor: {}\n", vm_path, major_id, minor_id
-    );
-
-    if let Some(idxdata) = devidx.0.get(host_path) {
-        let dev = &mut linux.devices[idxdata.idx];
-        let host_major = dev.major;
-        let host_minor = dev.minor;
-
-        dev.major = major_id as i64;
-        dev.minor = minor_id as i64;
-        dev.path = final_path.to_string();
-
-        info!(
-            sl!(),
-            "change the device from path: {} major: {} minor: {} to vm device path: {} major: {} minor: {}",
-            host_path,
-            host_major,
-            host_minor,
-            dev.path,
-            dev.major,
-            dev.minor,
-        );
+    if let Some(resources) = linux.resources.as_mut() {
+        for r in &mut resources.devices {
+            if let (Some(host_major), Some(host_minor)) = (r.major, r.minor) {
+                if let Some(update) = resupdates.get(&(r.r#type.as_str(), host_major, host_minor)) {
+                    info!(
+                        sl!(),
+                        "update_spec_devices() updating resource";
+                        "type" => &r.r#type,
+                        "host_major" => host_major,
+                        "host_minor" => host_minor,
+                        "guest_major" => update.guest_major,
+                        "guest_minor" => update.guest_minor,
+                    );
+
+                    r.major = Some(update.guest_major);
+                    r.minor = Some(update.guest_minor);
+                }
+            }
+        }
+    }
 
-        // Resources must be updated since they are used to identify
-        // the device in the devices cgroup.
-        for ridx in &idxdata.residx {
-            // unwrap is safe, because residx would be empty if there
-            // were no resources
-            let res = &mut linux.resources.as_mut().unwrap().devices[*ridx];
-            res.major = Some(major_id as i64);
-            res.minor = Some(minor_id as i64);
+    // Correct PCI addresses in the environment
+    if let Some(process) = spec.process.as_mut() {
```

Comment:
Again, this would shrink if you could move this block into a `set_pci_envvars()` function or similar.